### DISCUSSION TITLE
Decide on contribution workflow

### DISCUSSION THREAD

**Discussion by sjvrijn** [SS-NES/SS-NES-lesson#4] [General]
Here's my proposal for how we can define our collaborative workflow on the Lesson material development:

Short version:

0. Make an issue
1. Work on a dedicated branch
2. Create a PR
3. Suggest reviewer(s)
4. Have PR reviewed
5. When approved by >= 1 person, PR author can merge, otherwise latest reviewer

Specifically, I would suggest that only 1 approval is needed, and that the PR author can merge the branch themselves, assuming they have the permission to. Otherwise, the (last) reviewer can simply do the merge instead.

Feedback is appreciated, especially if you disagree. Otherwise, feel free to just mark your agreement with a quick emoji reaction 👍 ❤️ 🚀 

Once we reach agreement, I'll codify the workflow in a CONTRIBUTING.md

---

Long version:

0. **Make a (short) issue about the change or addition.**
   It can be as simple as just a 1-sentence summary in the title, but any relevant motivation or references are always helpful. This helps to separate the discussions on _what_ to do from _how_ to achieve it.
1. **Create a dedicated branch for your proposed change.**
   A common style of naming branches is e.g. `issue_number-issue_description`, e.g. `42-add-citation-section`.
2. **Create a Pull Request (PR) when the work is done or needs feedback/help.**
   The PR should refer back to any issue its addressing, and should contain a short description of what's been done. Any notes on particular decisions/trade-offs that have been made are also very relevant to include.
3. **Suggest reviewer(s).**
   On the right-hand side, you can assign some reviewers from the organization. This is a suggestion of one or two people who you think are relevant to review your work. If you have been suggested and do not have time to review, just let this be known in a comment and suggest someone else to take over.
4. **Have the PR reviewed.**
   Be nice 🙂 When you agree, mark the PR with your approval
5. **Once approved, author can merge.**
   When a PR has been approved by at least one person and there are no further active discussions, the original PR author can merge the PR. The author is also responsible for fixing any conflicts that may show up because other branches have been merged in the mean time. Any remaining active discussions should be made into separate issues before merging.
   If the PR author is not part of the organization and therefore does not have the permission to make the merge, then the last person to approve can make the merge instead.

**Comment by GiulioRosani** [comment]
I'm fine with this suggestion.

**Comment by teunoknmi** [comment]
How do we deal with the gh-pages branch since that's the branch that updates the [website](https://ss-nes.github.io/SS-NES-lesson/)?

**Reply by teunoknmi** [reply]
And we started with 2 people on this branch https://github.com/SS-NES/SS-NES-lesson/pull/3 . Do we merge this one first or cherrypick the changes to another branch?

**Reply by sjvrijn** [reply]
For the `gh-pages` branch, I guess it would be easiest to simply make that the main branch of the repo, and make all PRs directly to that branch so that the latest approved version is always shown.

For your other question, I would simply say that the PR author is always in the lead, and by default you would only make a PR from your own branch. If you've agreed somewhere else (either online or offline) to work together, then that's fine, and you can decide among yourselves if you want to distribute the responsibilities differently. But if the PR ends up 'stale', then the PR creator will be the one pinged for an update.

**Reply by teunoknmi** [reply]
Let's make that the main then branch then.

**Reply by sjvrijn** [reply]
~Done! I'll keep an eye on merging any remaining commits on the main branch into `gh-pages` and removing the main branch after the fact.~

**Reply by neutvd** [reply]
I just noticed that when creating an issue and then creating a branch from the issue page, that it was created from `gh-pages` branch. I had not read this discussion yet, so this was quite confusing. The `episodes/introduction.md` file is in the `main` branch, but not in the `gh-pages` branch, should we merge that into `gh-pages`? I've made a mess of my branches now, because I thought I had made a mistake and deleted the branch I created from the issue again and am about to commit a new one based on the `main` branch.

**Reply by sjvrijn** [reply]
Seems like I misunderstood the `gh-pages` branch: I think that's being updated automatically by the GitHub actions that run to publish the website. In any case, it's not helpful to work in that branch, since the original episodes folder is (automatically) deleted from there, so let's just stick to the `main` branch instead, and the `gh-pages` branch should follow on its own.

I've reverted the `main` branch to be our 'main' branch, so any new branches created from the issue pages should be correct again. Let's see how this goes and keep an eye on `gh-pages` to see what it does.
